app/src/main/java/com/github/savrov/research/compose/topic/state/SmartRecomposition.kt

```diff
@@ -0,0 +1,123 @@
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Row
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.lazy.LazyColumn
+import androidx.compose.foundation.lazy.items
+import androidx.compose.material3.Button
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import kotlinx.parcelize.Parcelize
+import kotlin.random.Random
+
+@Composable
+fun SmartRecomposition() {
+    var items by rememberSaveable { mutableStateOf(listOf<Item>()) }
+    val onAddAbove = {
+        items = listOf(Item(ItemId(items.size + 1))) + items
+    }
+    val onAddBelow = {
+        items = items + Item(ItemId(items.size + 1))
+    }
+    val onClear = { items = emptyList() }
+    SmartListComponent(items = items, onAddAbove = onAddAbove, onAddBelow = onAddBelow, onClear = onClear)
+}
+
+@Composable
+private fun SmartListComponent(
+    modifier: Modifier = Modifier,
+    items: List<Item>,
+    onAddAbove: () -> Unit,
+    onAddBelow: () -> Unit,
+    onClear: () -> Unit
+) {
+    LazyColumn(modifier = modifier.fillMaxSize()) {
+        item(key = "buttons") {
+            SmartListButtons(onAddAbove = onAddAbove, onAddBelow = onAddBelow, onClear = onClear)
+        }
+        items(items, key = { item -> item.id.value }) { item ->
+            SmartListItem(item = item)
+        }
+    }
+}
+
+@Composable
+private fun SmartListItem(
+    modifier: Modifier = Modifier,
+    item: Item
+) {
+    val backgroundColor = Color.random()
+    Row(
+        modifier = modifier
+            .fillMaxWidth()
+            .padding(16.dp)
+            .background(color = backgroundColor)
+            .padding(8.dp),
+        horizontalArrangement = Arrangement.Center
+    ) {
+        Text(text = "Item ${item.id}", color = Color.revert(backgroundColor))
+    }
+}
+
+@Composable
+private fun SmartListButtons(
+    modifier: Modifier = Modifier,
+    onAddAbove: () -> Unit,
+    onAddBelow: () -> Unit,
+    onClear: () -> Unit
+) {
+    Row(
+        modifier = modifier
+            .padding(16.dp)
+            .fillMaxSize(),
+        horizontalArrangement = Arrangement.SpaceBetween
+    ) {
+        Button(onClick = onAddAbove) {
+            Text(text = "Add above")
+        }
+        Button(onClick = onClear) {
+            Text(text = "Clear")
+        }
+        Button(onClick = onAddBelow) {
+            Text(text = "Add below")
+        }
+    }
+}
+
+@Parcelize
```

Comment:
I tried to test a @Stable annotation here, but did not see any difference in recomposition :/ Still do not understand when to use it.

Reply:
@Savrov 
regarding your question :

@Stable is like a special note in Jetpack Compose that tells us a composable or data type won't change much in how it works for users. This stability lets the application be clever about updating things, making it work faster and smarter.

You can use @Stable to improve situations where the information is organized in a complicated way, like when it's nested or involves groups of data.

in short : in most of the complex case we use @Stable

here is good article regarding @Stable  :
[@Stable ](https://medium.com/@android-world/optimizing-jetpack-compose-with-stable-smart-recomposition-strategies-81147b779a0a#:~:text=%40Stable%20is%20an%20annotation%20in,smarter%20and%20more%20efficient%20recompositions.)

please let me know if you have question